### PR TITLE
Verify JWT and use subject

### PR DESCRIPTION
**Pull request by leplatrem**
```yaml
# policies.yaml 
-
  id: "1"
  description: This policy allows Google users to do everything
  subjects:
    - google-oauth2|<.*>
  actions:
    - <.*>
  resources:
    - <.*>
  effect: allow
```

Then I query `/allowed` with a JWT token, the token is verified, and the subject is used to match policies.

I pass the `Auth0-Domain` and `Auth0-Audience` in the headers, to be able to validate tokens from different services.

```http
$ echo {} | http POST :8080/allowed "Auth0-Domain:minimal-demo-iam" "Auth0-Audience:http://minimal-demo-iam.localhost:8000" "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6Ik5EZzFOemczTlRFeVEwVTFNMEZCTnpCQlFqa3hOVVk1UTBVMU9USXpOalEzUXpVek5UWkRNQSJ9.eyJpc3MiOiJodHRwczovL21pbmltYWwtZGVtby1pYW0uYXV0aDAuY29tLyIsInN1YiI6Imdvb2dsZS1vYXV0aDJ8MTA0MTAyMzA2MTExMzUwNTc2NjI4IiwiYXVkIjpbImh0dHA6Ly9taW5pbWFsLWRlbW8taWFtLmxvY2FsaG9zdDo4MDAwIiwiaHR0cHM6Ly9taW5pbWFsLWRlbW8taWFtLmF1dGgwLmNvbS91c2VyaW5mbyJdLCJpYXQiOjE1MDY5NTkxOTgsImV4cCI6MTUwNjk2NjM5OCwiYXpwIjoiV1lSWXBKeVM1RG5EeXhMVFJWR0NRR0NXR28yS05RTE4iLCJzY29wZSI6Im9wZW5pZCBwcm9maWxlIn0.ifelc8pX6JQipcarBE1b9mSjfn5H-DhDZ2uT5a_0Z14-brSsIOS2f74_RQXy8n5RNh_JbpQ8qPYHIyAVRHbEPT1yNzvPM9VMF6roW9hMpinE1oSd2D3IMc9kEp03J0Rw2yz3u5R1tLrMwVATpqSFtrFGclmnEYT8__zg2XQL3VLv5mvt-9YC-FOVeq8tCSGTeP9CrXp4AoQMOEJ-1RfSdMiy8tcg_PK6Xsrj86dxDRBDoSeTFO99UUA3yyrc0tFFwIvjvDv-Y0l7FpVnplpi_QkhCmYOMAjORgT0NlEtR2fJq2RZMqJcVCLVlfidAxEOu1Vkx4dcwLnqnk2HcEGb6Q" 
HTTP/1.1 200 OK
Content-Length: 16
Content-Type: application/json; charset=utf-8
Date: Mon, 02 Oct 2017 16:58:41 GMT

{
    "allowed": true
}
```
I takes around ~200ms, but I guess `JWKClient` could be cached by domain... Later the authorization could be in charge of fetching LDAP userid and groups to match the policies (instead of just of the JWT subject)

@Natim @mozbhearsum @mostlygeek @peterbe Any thought about this POC? 
Thanks a lot!

### CODE REVIEW

**Comment by mozbhearsum** [comment]
It looks like this app as a whole is just a very, very thin wrapper around Warden? I guess I'm not sure what exactly you're looking for review on.

**Comment by mostlygeek** [comment]
The JWT's headers/payload (for reference) 

Header: 
````
{
  "typ": "JWT",
  "alg": "RS256",
  "kid": "NDg1Nzg3NTEyQ0U1M0FBNzBBQjkxNUY5Q0U1OTIzNjQ3QzUzNTZDMA"
}
````
Payload: 

````
{
  "iss": "https://minimal-demo-iam.auth0.com/",
  "sub": "google-oauth2|104102306111350576628",
  "aud": [
    "http://minimal-demo-iam.localhost:8000",
    "https://minimal-demo-iam.auth0.com/userinfo"
  ],
  "iat": 1506959198,
  "exp": 1506966398,
  "azp": "WYRYpJyS5DnDyxLTRVGCQGCWGo2KNQLN",
  "scope": "openid profile"
}
````

I don't like using custom headers to identify the issuer.  Instead use the issuer (`iss`) in the payload.  No `iss` == auto fail. 

If the algorithm is asymmetric (e.g RS256)  ... we should fetch / cache the public key to prevent more look ups.

**Comment by mostlygeek** [comment]
also committing ./vendor in a PR makes it really hard to review ... 
don't really have any good suggestions other than merging in after a PR.

**Comment by Natim** [comment]
Is it possible to consider not vendoring libs?

**Comment by Natim** [comment]
> It looks like this app as a whole is just a very, very thin wrapper around Warden?

@mozbhearsum just to make sure we are talking about the same thing, the content of the warden folder in this PR is the code of the project @leplatrem wrote, not sure if you were refering to the ory/warden project here.

File: warden/jwt.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	"net/http"
+
+	auth0 "github.com/auth0-community/go-auth0"
+	"github.com/gin-gonic/gin"
+	jose "gopkg.in/square/go-jose.v2"
+	jwt "gopkg.in/square/go-jose.v2/jwt"
+)
+
+func verifyJWT(request *http.Request) (*jwt.Claims, error) {
+	domain := request.Header.Get("Auth0-Domain")
```
Comment (Natim) [inline warden/jwt.go]: Should this be part of the config of the instance? Probably read from the ENV variables?

File: warden/jwt.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+		return nil, errors.New("Auth0-Domain header missing")
+	}
+
+	apiIdentifer := request.Header.Get("Auth0-Audience")
+	if apiIdentifer == "" {
+		return nil, errors.New("Auth0-API-Identifier header missing")
+	}
+
+	jwksURI := fmt.Sprintf("https://%s.auth0.com/.well-known/jwks.json", domain)
+	apiIssuer := fmt.Sprintf("https://%s.auth0.com/", domain)
```
Comment (Natim) [inline warden/jwt.go]: I don't think this should be specific to auth0 but most likely specific to OpenID Connect IDP, we should probably also use the config to store the domain including ``auth0.com``

File: warden/warden.go [hunk 3]
```diff
@@ -11,6 +11,7 @@ import (
 	"github.com/ory/ladon"
 	manager "github.com/ory/ladon/manager/memory"
 	log "github.com/sirupsen/logrus"
+	jwt "gopkg.in/square/go-jose.v2/jwt"
```
Comment (Natim) [inline warden/warden.go]: 👍

File: warden/jwt.go (hunk 1 above)
Comment (leplatrem) [reply warden/jwt.go]: Maybe for this one indeed, but not for the audience right? (each Relying Party like Balrog have their own audience, unless I'm mistaken...)

**Comment by leplatrem** [comment]
> Is it possible to consider not vendoring libs? 

I guess that since we use `glide`, there should be something possible...

File: warden/jwt.go (hunk 1 above)
Comment (Natim) [reply warden/jwt.go]: Yes then would we be able to grab it from the referer?

Comment (leplatrem) [reply warden/jwt.go]: The caller (Relying Party, like balrog) is not a browser, so it has to send it explicitly

**Comment by leplatrem** [comment]
This branch was demoed. No one seemed to react with a funny face, so I'll add some tests and merge :)

**Pull request merged**